### PR TITLE
Fix query paramater encoding when rewriting proxied requests.

### PR DESCRIPTION
**Pull request by olix0r**
The DiscardProxySideEffects module double-encodes URIs, so that a query param
like `text=hello%20world` is encoded as `text=hello%2520world`. Fix rewritten
URI construction to prevent double-encoding.

Several other minor changes are included:

- rename DiscardProxySideEffects to ProxyRewriteFilter (for clarity)
- make filter static -- there's no benefit to instantiating new filters in this scenario
- only remove 3 well-known proxy headers -- since header stripping runs on ALL
  requests, we should aim to make it as cheap as possible.

### CODE REVIEW

File: router/http/src/main/scala/io/buoyant/router/http/ProxyRewriteFilter.scala [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      req.uri = unproxifyUri(uri)
+    }
+  }
+
+  /**
+   * Return only the path, query, and fragment segments of a URI.
+   */
+  private def unproxifyUri(orig: URI): String =
+    if (!orig.isAbsolute) orig.toString
+    else new URI(null, null, orig.getPath, orig.getQuery, orig.getFragment).toString
```
Comment (olix0r) [inline router/http/src/main/scala/io/buoyant/router/http/ProxyRewriteFilter.scala]: This is the main meat of the fix, fwiw

File: router/http/src/main/scala/io/buoyant/router/http/DiscardProxySideEffects.scala [hunk 2]
```diff
@@ -1,39 +0,0 @@
...
-   * Strips host from request URI and drops Proxy-* headers
-   * Those side-effects are added by proxy-aware clients
-   */
-  class DiscardProxySideEffects extends SimpleFilter[Request, Response] {
-    def apply(req: Request, svc: Service[Request, Response]) = {
-      val uri = new URI(req.uri)
-
-      if (uri.isAbsolute) { // Proxied request
-        // TODO: HTTP/2.0 support once finagle supports it
-        val relativeUri = new URI(null, null, uri.getRawPath, uri.getRawQuery, uri.getRawFragment)
```
Comment (olix0r) [inline router/http/src/main/scala/io/buoyant/router/http/DiscardProxySideEffects.scala]: herein lies the bug

File: router/http/src/test/scala/io/buoyant/router/http/ProxyRewriteFiltertest.scala [hunk 3]
```diff
@@ -17,48 +17,44 @@ class DiscardProxySideEffectsTest extends FunSuite with Awaits {
...
 
-  test("filter drop Proxy-* Headers even for requests that doesn't look like proxied") {
+  test("strip Proxy-* Headers for non-proxy requests") {
     val req = Request(Method.Get, "/foo;matrix=uri;/bar?x=1&y=2")
     req.host = "acme.co:8080"
     req.headerMap.set("Proxy-Connection", "Keep-Alive")
     req.headerMap.set("Proxy-Authorization", "Basic QWxhZGRpbjpvcGVuIHNlc2FtZQ==")
     req.headerMap.set("Proxy-Authenticate", "Basic")
-    req.headerMap.set("Proxy-Foo", "Bar")
+    // req.headerMap.set("Proxy-Foo", "Bar")
```
Comment (adleong) [inline router/http/src/test/scala/io/buoyant/router/http/ProxyRewriteFiltertest.scala]: rm

File: router/http/src/test/scala/io/buoyant/router/http/ProxyRewriteFiltertest.scala [hunk 4]
```diff
@@ -17,48 +17,44 @@ class DiscardProxySideEffectsTest extends FunSuite with Awaits {
...
-  test("filter rewrites uri and drops Proxy-* headers for proxied requests") {
-    val req = Request(Method.Get, "http://acme.co:8080/foo;matrix=uri;/bar?x=1&y=2")
+  test("rewrite proxy requests as non-proxy requests") {
+    val req = Request(Method.Get, "http://acme.co:8080/foo;matrix=uri;/bar?x=1&y=2&z=%28%20hi%20%29")
     req.host = "example.com:9000"
     req.headerMap.set("Proxy-Connection", "Keep-Alive")
     req.headerMap.set("Proxy-Authorization", "Basic QWxhZGRpbjpvcGVuIHNlc2FtZQ==")
     req.headerMap.set("Proxy-Authenticate", "Basic")
-    req.headerMap.set("Proxy-Foo", "Bar")
+    // req.headerMap.set("Proxy-Foo", "Bar")
```
Comment (adleong) [inline router/http/src/test/scala/io/buoyant/router/http/ProxyRewriteFiltertest.scala]: rm

**Comment by adleong** [comment]
:star: 👸 🐂 📧

**Pull request merged**